nextstrain/cli/remote/s3.py

```diff
@@ -78,33 +43,128 @@ def upload(local_files: List[Path], bucket, prefix: str) -> List[str]:
     files = list(zip(local_files, [ prefix + f.name for f in local_files ]))
 
     for local_file, remote_file in files:
-        print("Deploying", local_file, "as", remote_file)
+        yield local_file, Path(remote_file)
 
         # Upload compressed data
-        with local_file.open("rb") as data, gzip_stream(data) as gzdata:
+        with GzipCompressingReader(local_file.open("rb")) as gzdata:
             bucket.upload_fileobj(
                 gzdata,
                 remote_file,
                 { "ContentType": content_type(local_file), "ContentEncoding": "gzip" })
 
-    return [ remote for local, remote in files ]
+    # Purge any CloudFront caches for this bucket
+    purge_cloudfront(bucket, [remote for local, remote in files])
+
+
+def download(url: urllib.parse.ParseResult, local_path: Path, recursively: bool = False) -> Iterable[Tuple[Path, Path]]:
+    """
+    Download the files deployed at the given remote *url*, optionally
+    *recursively*, saving them into the *local_dir*.
+    """
+    bucket, path = split_url(url)
+
+    # Download either all objects sharing a prefix or the sole object (if any)
+    # with the given key.
+    if recursively:
+        objects = [ item.Object() for item in bucket.objects.filter(Prefix = path) ]
+    else:
+        objects = [ bucket.Object(path) ]
+
+    def local_file_path(obj):
+        if local_path.is_dir():
+            return local_path / Path(obj.key).name
+        else:
+            return local_path
+
+    files = list(zip(objects, [local_file_path(obj) for obj in objects]))
+
+    for remote_object, local_file in files:
+        yield Path(remote_object.key), local_file
+
+        encoding = remote_object.content_encoding
 
+        with ContentDecodingWriter(encoding, local_file.open("wb")) as file:
+            remote_object.download_fileobj(file)
 
-def gzip_stream(stream):
+
+def ls(url: urllib.parse.ParseResult) -> Iterable[Path]:
+    """
+    List the files deployed at the given remote *url*.
+    """
+    bucket, prefix = split_url(url)
+
+    return [ Path(obj.key) for obj in bucket.objects.filter(Prefix = prefix) ]
+
+
+def delete(url: urllib.parse.ParseResult, recursively: bool = False) -> Tuple[Iterable[str], Iterable[Tuple[str, ...]]]:
+    """
+    Delete the files deployed at the given remote *url*, optionally *recursively*.
+    """
+    bucket, path = split_url(url)
+
+    # Delete either all objects sharing a prefix or the sole object (if any)
+    # with the given key.  Both methods return the same results data structure.
+    objects = bucket.objects.filter(Prefix = path)
+
+    if recursively:
+        responses = objects.delete()
```

Comment:
I haven't tested this, but it looks as if this will allow (recursive) deletion of an entire bucket. I can see this being useful in certain cases (groups) but also very dangerous. Please consider requiring `path` to be set for recursive deletion.